src/objc/AudioPreprocessing.m

```diff
@@ -0,0 +1,365 @@
+/* Copyright © 2019 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#include <math.h>
+
+#import <Accelerate/Accelerate.h>
+
+#import "AudioPreprocessing.h"
+
+
+// TODO: Make some of these parameters to the model
+const int inputLength = 15600;
+const NSUInteger windowLength = 400;
+const int hopLength = 160;
+
+const float minFrequency = 125;
+const float maxFrequency = 7500;
+
+const NSUInteger numBands = 64;
+const double sampleRate = 16000;
+const NSUInteger fftLength = 512;
+const int spectrumSize = 257;
+
+const float log_offset = 0.01;
+
+
+API_AVAILABLE(macos(10.13))
+static void setErrorMsg(const char* errorMsg,  NSError** error) {
+    if(error) {
+      *error = [NSError errorWithDomain:MLModelErrorDomain
+                                   code:MLModelErrorGeneric
+                               userInfo:@{NSLocalizedDescriptionKey: @(*errorMsg)}];
+    }
+}
+
+static inline float hzToMel(float freq) {
+    return 1127.0 * log(1 + freq/700.0f);
+}
+
+
+@interface PowerSpectrum : NSObject
+
+@property (readonly) float* PowerSpectrum;
+
+-(id)initWithFFTLength:(NSUInteger)fftLength withFrameLength:(NSUInteger)frameLength;
+-(void)applyHammingWindow:(float*)signal numSamples:(NSUInteger)numSamples;
+-(void)calculatePowerSpectrum:(float*)signal numSamples:(NSUInteger)numSamples;
+
+@end
+
+
+@implementation PowerSpectrum {
+    float *_hammingWindow, *_workingFrame;
+
+    DSPSplitComplex _spectrum;
+    FFTSetup _fftSetup;
+    
+    NSUInteger _frameLength, _fftLength;
+}
+
+
+-(id)initWithFFTLength:(NSUInteger)fftLength withFrameLength:(NSUInteger)frameLength {
+    
+    self = [super init];
+    
+    _fftLength = fftLength;
+    _frameLength = frameLength;
+    
+    _hammingWindow = (float*)calloc(frameLength, sizeof(float));
+    [self _generateHammingWindow:frameLength];
+    
+    _fftSetup = vDSP_create_fftsetup(log2(fftLength*2), FFT_RADIX2);
+    _spectrum.realp = (float*)calloc(fftLength, sizeof(float));
+    _spectrum.imagp = (float*)calloc(fftLength, sizeof(float));
+    _PowerSpectrum = (float*)calloc(fftLength/2 + 1, sizeof(float));
```

Comment:
Strange, I don't understand where `_PowerSpectrum` is declared

Reply:
Yeah, as you mentioned in a later comment, I think it was actually a global. That's fixed now.